pkg/driver/kubernetes.go

```diff
@@ -0,0 +1,372 @@
+	pods                  coreclientv1.PodInterface
+	deletionPolicy        metav1.DeletionPropagation
+	requiredCompletions   int32
+}
+
+// NewKubernetesDriver initializes a Kubernetes driver.
+func NewKubernetesDriver(namespace, serviceAccount string, conf *rest.Config) (*KubernetesDriver, error) {
+	driver := &KubernetesDriver{
+		Namespace:          namespace,
+		ServiceAccountName: serviceAccount,
+	}
+	driver.setDefaults()
+	err := driver.setClient(conf)
+	return driver, err
+}
+
+// Handles receives an ImageType* and answers whether this driver supports that type.
+func (k *KubernetesDriver) Handles(imagetype string) bool {
+	return imagetype == driver.ImageTypeDocker || imagetype == driver.ImageTypeOCI
+}
+
+// Config returns the Kubernetes driver configuration options.
+func (k *KubernetesDriver) Config() map[string]string {
+	return map[string]string{
+		"KUBE_NAMESPACE":  "Kubernetes namespace in which to run the invocation image",
+		"SERVICE_ACCOUNT": "Kubernetes service account to be mounted by the invocation image (if empty, no service account token will be mounted)",
+		"KUBE_CONFIG":     "Absolute path to the kubeconfig file",
+		"MASTER_URL":      "Kubernetes master endpoint",
+	}
+}
+
+// SetConfig sets Kubernetes driver configuration.
+func (k *KubernetesDriver) SetConfig(settings map[string]string) {
+	k.setDefaults()
+	k.Namespace = settings["KUBE_NAMESPACE"]
+	k.ServiceAccountName = settings["SERVICE_ACCOUNT"]
+
+	var kubeconfig string
+	if kpath := settings["KUBE_CONFIG"]; kpath != "" {
+		kubeconfig = kpath
+	} else if home := homeDir(); home != "" {
+		kubeconfig = filepath.Join(home, ".kube", "config")
+	}
+
+	conf, err := clientcmd.BuildConfigFromFlags(settings["MASTER_URL"], kubeconfig)
+	if err != nil {
+		panic(err)
+	}
+	err = k.setClient(conf)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (k *KubernetesDriver) setDefaults() {
+	k.SkipCleanup = false
+	k.BackoffLimit = 0
+	k.ActiveDeadlineSeconds = 300
+	k.requiredCompletions = 1
+	k.deletionPolicy = metav1.DeletePropagationBackground
+}
+
+func (k *KubernetesDriver) setClient(conf *rest.Config) error {
+	coreClient, err := coreclientv1.NewForConfig(conf)
+	if err != nil {
+		return err
+	}
+	batchClient, err := batchclientv1.NewForConfig(conf)
+	if err != nil {
+		return err
+	}
+	k.jobs = batchClient.Jobs(k.Namespace)
+	k.secrets = coreClient.Secrets(k.Namespace)
+	k.pods = coreClient.Pods(k.Namespace)
+
+	return nil
+}
+
+// Run executes the operation inside of the invocation image.
+func (k *KubernetesDriver) Run(op *driver.Operation) error {
+	if k.Namespace == "" {
+		return fmt.Errorf("KUBE_NAMESPACE is required")
+	}
+	labelMap := generateLabels(op)
+	meta := metav1.ObjectMeta{
+		Namespace:    k.Namespace,
+		GenerateName: generateNameTemplate(op),
+		Labels:       labelMap,
+	}
+	// Mount SA token if a non-zero value for ServiceAccountName has been specified
+	mountServiceAccountToken := k.ServiceAccountName != ""
+	job := &batchv1.Job{
```

Comment:
@sbawaska demonstrated this PR to our team today and it was very well received. However, it was pointed out that in some circumstances jobs can restart even when you think you have configured them not to. I raised https://github.com/deislabs/duffle/issues/771 to cover this. Feel free to implement this issue in this PR or merge and we can implement it later.

Reply:
Actually, upon further investigation, it seems that the situation when jobs restart when we don't want them to may have been fixed in k8s, so I've closed https://github.com/deislabs/duffle/issues/771. Sorry for the false alarm.